Common/src/main/java/com/blamejared/crafttweaker/api/util/ParseUtil.java

```diff
@@ -1,54 +1,87 @@
+import org.openzen.zenscript.parser.expression.ParsedCallArguments;
 import org.openzen.zenscript.parser.expression.ParsedExpression;
 import org.openzen.zenscript.parser.expression.ParsedExpressionMember;
+import org.openzen.zenscript.parser.expression.ParsedExpressionString;
 import org.openzen.zenscript.parser.expression.ParsedExpressionVariable;
+import org.openzen.zenscript.parser.expression.ParsedNewExpression;
 import org.openzen.zenscript.parser.type.IParsedType;
 import org.openzen.zenscript.parser.type.ParsedNamedType;
 
-import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.List;
-import java.util.stream.Collectors;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public final class ParseUtil {
-
+    
+    private static final Pattern typeArgumentPattern = Pattern.compile("(.*)<(.*)>");
+    
     private ParseUtil() {
+    
     }
-
+    
     public static ParsedExpression explode(CodePosition position, String name) {
-
+        
         String[] strExpressions = name.split("\\.");
         ParsedExpression expression = new ParsedExpressionVariable(position, strExpressions[0], null);
-        if (strExpressions.length > 1) {
-            for (int i = 1; i < strExpressions.length; i++) {
+        if(strExpressions.length > 1) {
+            for(int i = 1; i < strExpressions.length; i++) {
                 expression = new ParsedExpressionMember(position, expression, strExpressions[i], List.of());
             }
         }
-
+        
         return expression;
     }
-
+    
     public static IParsedType readParsedType(String name, CodePosition position) {
-
-        final List<ParsedNamedType.ParsedNamePart> collect = Arrays.stream(name.split("[.]"))
-                .map(s -> new ParsedNamedType.ParsedNamePart(s, null))
-                .collect(Collectors.toList());
-
-        return new ParsedNamedType(position, collect);
+        
+        List<ParsedNamedType.ParsedNamePart> parsedParts = new LinkedList<>();
+        while(!name.isBlank()) {
+            int end = name.length();
+            if(name.contains(".")) {
+                end = name.indexOf(".");
+            }
+            String namePart = name.substring(0, end);
+            if(namePart.contains("<") && name.contains(">")) {
+                end = name.lastIndexOf(">") + 1;
+                namePart = name.substring(0, name.lastIndexOf(">") + 1);
+            }
+            name = name.substring(Math.min(name.length(), end + 1));
+            
+            Matcher matcher = typeArgumentPattern.matcher(namePart);
+            if(matcher.matches()) {
+                parsedParts.add(new ParsedNamedType.ParsedNamePart(matcher.group(1), List.of(readParsedType(matcher.group(2), position))));
+            } else {
+                parsedParts.add(new ParsedNamedType.ParsedNamePart(namePart, null));
+            }
+        }
+        
+        return new ParsedNamedType(position, parsedParts);
```

Comment:
Can you verify this works even with nested generics like `stdlib.List<stdlib.List<string>[]>`?